Rework timing info for --profile-rule-parsing

radumereuta opened the pull request:
This changes a bit the way `--profile-rule-parsing` generates timing.log files:
- print the collected messages into a single file specified at command line
- changed the timings from seconds with decimals to milliseconds (since I find them more readable)
- move the option to OuterParsingOptions to make it available in other tools as well

Here is a sample from the output file:
```
   18 /home/radu/work/k/k-distribution/include/kframework/builtin/domains.md:868   parse:   2 typeInf:   7
    9 /home/radu/work/k/k-distribution/include/kframework/builtin/domains.md:869   parse:   2 typeInf:   0
   28 /home/radu/work/k/k-distribution/include/kframework/builtin/domains.md:870   parse:  10 typeInf:   5
   16 /home/radu/work/k/k-distribution/include/kframework/builtin/domains.md:872   parse:   2 typeInf:   0
   38 /home/radu/work/k/k-distribution/include/kframework/builtin/kast.md:523   parse:  12 typeInf:   9
   38 /home/radu/work/test/test.k:11   parse:  13 typeInf:  10
   38 /home/radu/work/test/test.k:5   parse:  12 typeInf:  10
```
All times in milliseconds.
The first column is total parsing time aligned to the right. If something is way bigger, it's going to pop out immediately.
The entries are ordered by the second column to provide a stable output. It's easy to compare different runs.
At the end, I added split timings for parsing time and type inference. If parsing time is large, you may want to add parentheses, if the type inference is large, then you may want to type your variables. If I find more relevant information, I can easily add it later.

hreada commented:
Created a clean directory for matching logic prover module.

grosu commented:
Cool!



Sent from my Verizon, Samsung Galaxy smartphone


-------- Original message --------
From: Xiaohong Chen <notifications@github.com>
Date: 9/25/16 16:40 (GMT-06:00)
To: kframework/k <k@noreply.github.com>
Subject: Re: [kframework/k] Add new repository "/prover" for matching logic prover. (#2204)


Created a clean directory for matching logic prover module.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_kframework_k_pull_2204-23issuecomment-2D249448399&d=CwMCaQ&c=8hUWFZcy2Z-Za5rBPlktOQ&r=gBub4ab7wvmjFVEDwsswdud0fptTHrGi3G4Zzc5Kwds&m=FsnBUNMtb7rAS0DbQ9jMkwktrf43AFC7KB-Q7HUCgho&s=PbxWkiSWrbizgkxuO5qQC6jFvApPaquczsA8dMZ9OYc&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ADj44f8fcYAWD6awhXlcxY5-5FFsXHCwPUks5qtupNgaJpZM4KGAzL&d=CwMCaQ&c=8hUWFZcy2Z-Za5rBPlktOQ&r=gBub4ab7wvmjFVEDwsswdud0fptTHrGi3G4Zzc5Kwds&m=FsnBUNMtb7rAS0DbQ9jMkwktrf43AFC7KB-Q7HUCgho&s=67B9vJ6KteAhzEQwxoWPHW2JgREvdJN5uUXfatgLSJQ&e=>.

radumereuta committed:
Move --profile--rules to OuterParsingOptions

radumereuta committed:
Rework timing info

- print to temp dir timings from each thread
- collect everything into a single file specified at cmd
- add test

dwightguth commented on k-distribution/tests/regression-new/profile/Makefile:
```diff
@@ -1,5 +1,11 @@
 DEF=test
 EXT=test
-KOMPILE_FLAGS=--profile-rule-parsing --syntax-module TEST
+KOMPILE_BACKEND=java
```
Why was the backend of this test changed? In general we want to avoid adding new tests that run on the ocaml or java backends when possible, unless we are specifically testing those backends.

radumereuta replied:
That's fair. I needed to test kprove besides kompile. I will change it to haskell.

dwightguth commented on kernel/src/main/java/org/kframework/utils/options/OuterParsingOptions.java:
```diff
@@ -51,4 +51,7 @@ public synchronized File mainDefinitionFile(FileUtil files) {
     public String mdSelector = "k";
+
+    @Parameter(names="--profile-rule-parsing", description="Generate time in seconds to parse each rule in the semantics. Found in -kompiled directory under timing.log.")
+    public String profileRules;
```
This is not an appropriate place for this flag because it does not relate to outer parsing.

radumereuta replied:
The only alternative would be to make an entirely new `InnerParsingOptions.java` class and add it to kompile and kprove (not sure if it's needed anywhere else). Is this what you have in mind?

dwightguth replied:
that would probably be better. Right now, this flag is being included in KDep, which is completely inappropriate.

dwightguth commented on kernel/src/main/java/org/kframework/parser/inner/ParseInModule.java:
```diff
@@ -228,10 +228,7 @@ public Scanner getScanner() {
-        if (profileRules) {
-            start = System.nanoTime();
-        }
+        long start = System.currentTimeMillis(), endParse = 0, startTypeInf = 0, endTypeInf = 0;
```
I don't understand why this is being changed. Not only are we reducing the precision by a factor of 1000, you are now also calling the potentially somewhat expensive function even when its value doesn't matter.

radumereuta replied:
According to this article: https://www.geeksforgeeks.org/java-system-nanotime-vs-system-currenttimemillis/
`System.nanoTime()`
- may not be thread safe (although I've read that might not be the case with modern java versions)
- `currentTimeMillis()` executes faster than `nanoTime()` 
- the precision is irrelevant. I did some tests with nanotime and the time variation for the same rule can be 30ms. Considering that we are running multi threaded, cache misses are non-deterministic, opportunistic CPU boost... this variation is understandable. So milliseconds is completely fine in our case. We are looking for big outliers (100ms or more).

I don't feel strongly about it, I can change it back to `nanoTime()`. But according to my research, `currentTimeMillis()` is a better fit for our case.

dwightguth replied:
Alright, this is fine then.

dwightguth replied:
We should still only call the function if we are actually profiling though.

dwightguth commented on kernel/src/main/java/org/kframework/parser/inner/ParseInModule.java:
```diff
@@ -248,6 +245,7 @@ public Scanner getScanner() {
             } catch (KEMException e) {
                 return Tuple2.apply(Left.apply(Collections.singleton(e)), Collections.emptySet());
             }
+            endParse = System.currentTimeMillis();
```
Same issue throughout here

dwightguth commented on kernel/src/main/java/org/kframework/parser/inner/ParseInModule.java:
```diff
@@ -300,18 +300,15 @@ public Scanner getScanner() {
-                        t.write(' ');
-                        t.write(Double.toString((stop - start) / 1000000000.0));
-                        t.write('\n');
+                        long stop = System.currentTimeMillis();
```
By moving the stop time measurement inside the lock, you have added the totally irrelevant time spent waiting on the lock to the measurement.

radumereuta replied:
That's fair. BUT, that's also confusing. On #2116 you argued that using a `ConcurrentLinkedDeque` (which is thread-safe but contentious when adding elements) is not ok. But you are using a lock here.

What is going on? How is this code working/supposed to work?

dwightguth replied:
You're right that something weird is going on. When I first wrote this code, each thread had its own ParseInModule instance so it didn't really matter that there was a lock there. At some point that changed so each thread uses the same ParseInModule instance, and we didn't address this inconsistency.  You probably want to change the writer so it's inside a ThreadLocal, I think, and remove the lock.

radumereuta replied:
I see 4 possible solutions here:
1. Use ThreadLocal to print the timing info in a file per thread. First google link I found though warns about using ThreadLocal+ThreadPools: https://www.baeldung.com/java-threadlocal#threadlocalsand-thread-pools
2. Generate a new tempfile for each parsed rule. Basically, bypass the synchronization problem entirely. We're combining them into a single file anyway. But we create more IO traffic.
3. Use a `ConcurrentLinkedQueue` which (as I said above) is threadsafe but contentious when adding elements. But the implementation is based on volatility ([Compare-And-Swap](https://www.baeldung.com/java-queue-linkedblocking-concurrentlinked#concurrentlinkedqueue)). Meaning the variable it's writing into is _not_ stored in cache, only in the main memory, and it offers very low overhead and low probability of collision. Since we're doing this operation outside of the area we're measuring, it's fine. I see this solution as a lower cost than writing to disk.
4. `Tuple2<Errors, Term> parseInModule` becomes `Tuple3<Errors, Term, Timing> parseInModule`. In other words, bypass the thread synchronization issue, and let the stream collectors put everything together. I may even be able to add timing info for rules that fail to parse.

dwightguth replied:
I like the second solution. We are only appending data so the I/O overhead should be pretty low since it will be sequential on disk. We can explore a different solution if it seems that the data changes too much based on this.

dwightguth reviewed (changes requested):
Overall, the logic to collate the data seems okay, but there are serious issues with the modifications you made to how the data is measured, and there's other problems with how the options parsing was modified.

radumereuta committed:
Print timing for each rule in separate tempfiles

then collect them into a single one after parsing

radumereuta committed:
Add InnerParsingOptions

And inject it everywhere

radumereuta committed:
Add option to other tools

rv-jenkins committed:
Merge branch 'master' into profileRules
